File: src/test/java/com/zepben/cimbend/network/tracing/AssignToFeedersTest.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.cimbend.network.tracing;

import com.zepben.cimbend.cim.iec61970.base.core.Feeder;
import com.zepben.cimbend.network.NetworkService;
import com.zepben.cimbend.testdata.PowerTransformersWithEndsNetwork;
import com.zepben.testutils.junit.SystemLogExtension;
import kotlin.jvm.JvmField;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.RegisterExtension;

import java.util.ArrayList;
import java.util.List;

import static com.zepben.cimbend.testdata.TestNetworks.getNetwork;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.containsInAnyOrder;

@SuppressWarnings("ConstantConditions")
public class AssignToFeedersTest {

    @JvmField
    @RegisterExtension
    public SystemLogExtension systemOutRule = SystemLogExtension.SYSTEM_ERR.captureLog().muteOnSuccess();

    @Test
    public void canRunFromFeederCb() {
        NetworkService network = getNetwork(7);

        Tracing.assignEquipmentContainersToFeeders().run(network);

        Feeder feeder = network.get(Feeder.class, "f");

        List<String> mRIDs = new ArrayList<>();
        feeder.getCurrentEquipment().forEach(equipment -> mRIDs.add(equipment.getMRID()));
        assertThat(mRIDs, containsInAnyOrder("fcb", "c2", "n1"));
    }

    @Test
    public void appliesWithTransformersWithEndsWithBaseVoltage() {
        NetworkService network = PowerTransformersWithEndsNetwork.createWithBaseVoltage();

        Tracing.assignEquipmentContainersToFeeders().run(network);

        Feeder feeder = network.get(Feeder.class, "f");

        List<String> mRIDs = new ArrayList<>();
        feeder.getCurrentEquipment().forEach(equipment -> mRIDs.add(equipment.getMRID()));
        assertThat(mRIDs, containsInAnyOrder("ztx", "c1", "fsp", "c2", "reg", "c3", "iso", "c4", "tx", "c5"));
    }

    @Test
    public void appliesWithTransformersWithEndsWithRatedVoltage() {
        NetworkService network = PowerTransformersWithEndsNetwork.createWithRatedVoltage();

        Tracing.assignEquipmentContainersToFeeders().run(network);

        Feeder feeder = network.get(Feeder.class, "f");

        List<String> mRIDs = new ArrayList<>();
        feeder.getCurrentEquipment().forEach(equipment -> mRIDs.add(equipment.getMRID()));
        assertThat(mRIDs, containsInAnyOrder("ztx", "c1", "fsp", "c2", "reg", "c3", "iso", "c4", "tx", "c5"));
    }

}

File: src/test/java/com/zepben/cimbend/network/tracing/AssignToFeedersTest.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.cimbend.network.tracing

import com.zepben.cimbend.cim.iec61970.base.core.Equipment
import com.zepben.cimbend.cim.iec61970.base.core.Feeder
import com.zepben.cimbend.testdata.*
import com.zepben.testutils.junit.SystemLogExtension
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.containsInAnyOrder
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.RegisterExtension

class AssignToFeedersTest {

    @JvmField
    @RegisterExtension
    var systemErr: SystemLogExtension = SystemLogExtension.SYSTEM_ERR.captureLog().muteOnSuccess()

    @Test
    fun appliesToEquipmentOnHeadTerminalSide() {
        val network = FeederStartPointBetweenConductorsNetwork.create()
        val feeder: Feeder = network["f"]!!

        Tracing.assignEquipmentContainersToFeeders().run(network)

        validateEquipment(feeder.equipment, "fsp", "c2")
    }

    @Test
    fun stopsAtNormallyOpenPoints() {
        val network = FeederStartPointToOpenPointNetwork.create(normallyOpen = true, currentlyOpen = false)
        val feeder: Feeder = network["f"]!!

        Tracing.assignEquipmentContainersToFeeders().run(network)

        validateEquipment(feeder.equipment, "fsp", "c1", "op")
        validateEquipment(feeder.currentEquipment, "fsp", "c1", "op", "c2")
    }

    @Test
    fun stopsAtCurrentlyOpenPoints() {
        val network = FeederStartPointToOpenPointNetwork.create(normallyOpen = false, currentlyOpen = true)
        val feeder: Feeder = network["f"]!!

        Tracing.assignEquipmentContainersToFeeders().run(network)

        validateEquipment(feeder.equipment, "fsp", "c1", "op", "c2")
        validateEquipment(feeder.currentEquipment, "fsp", "c1", "op")
    }

    @Test
    fun stopsAtSubstationTransformers() {
        val network = FeederToSubstationTransformerNetwork.create()
        val feeder: Feeder = network["f"]!!

        Tracing.assignEquipmentContainersToFeeders().run(network)

        validateEquipment(feeder.equipment, "fsp", "c1")
    }

    @Test
    fun stopsAtFeederStartPoints() {
        val network = DownstreamFeederStartPointNetwork.create(1)
        val feeder1: Feeder = network["f1"]!!
        val feeder2: Feeder = network["f2"]!!

        Tracing.assignEquipmentContainersToFeeders().run(network)

        validateEquipment(feeder1.equipment, "fsp1", "c2", "fsp2")
        validateEquipment(feeder2.equipment, "fsp1", "c2", "fsp2")
    }

    @Test
    fun stopsAtFeederStartPointsReversedHeadTerminal() {
        val network = DownstreamFeederStartPointNetwork.create(2)
        val feeder1: Feeder = network["f1"]!!
        val feeder2: Feeder = network["f2"]!!

        Tracing.assignEquipmentContainersToFeeders().run(network)

        validateEquipment(feeder1.equipment, "fsp1", "c2", "fsp2")
        validateEquipment(feeder2.equipment, "fsp2", "c3")
    }

    @Test
    fun handlesDroppedPhases() {
        val network = DroppedPhasesNetwork.create()
        val feeder: Feeder = network["f"]!!

        Tracing.assignEquipmentContainersToFeeders().run(network)

        validateEquipment(feeder.equipment, "fcb", "acls1", "acls2", "acls3", "iso", "acls4", "tx")
    }

    private fun validateEquipment(equipment: Collection<Equipment>, vararg expectedMRIDs: String) {
        assertThat(equipment.map { it.mRID }.toList(), containsInAnyOrder(*expectedMRIDs))
    }

}
